include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -261,12 +261,94 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
 
 		const CIncludeFinder* getDefaultIncludeFinder() const { return m_defaultIncludeFinder.get(); }
 
+		//string to be replaced with all "#" except those in "#include"
+		static constexpr const char* PREPROC_DIRECTIVE_DISABLER = "_this_is_a_hash_";
+		static constexpr const char* PREPROC_DIRECTIVE_ENABLER = PREPROC_DIRECTIVE_DISABLER;
+		static constexpr const char* PREPROC_GL__DISABLER = "_this_is_a_GL__prefix_";
+		static constexpr const char* PREPROC_GL__ENABLER = PREPROC_GL__DISABLER;
+		static constexpr const char* PREPROC_LINE_CONTINUATION_DISABLER = "_this_is_a_line_continuation_\n";
+		static constexpr const char* PREPROC_LINE_CONTINUATION_ENABLER = "_this_is_a_line_continuation_";
+
+		static void disableAllDirectivesExceptIncludes(std::string& _glslCode)
+		{
+			// TODO: replace this with a proper-ish proprocessor and includer one day
+			std::regex directive("#(?!(include|version|pragma shader_stage|line))");//all # not followed by "include" nor "version" nor "pragma shader_stage"
+			//`#pragma shader_stage(...)` is needed for determining shader stage when `_stage` param of IShaderCompiler functions is set to ESS_UNKNOWN
+			auto result = std::regex_replace(_glslCode, directive, PREPROC_DIRECTIVE_DISABLER);
+			std::regex glMacro("[ \t\r\n\v\f]GL_");
+			result = std::regex_replace(result, glMacro, PREPROC_GL__DISABLER);
+			std::regex lineContinuation("\\\\[ \t\r\n\v\f]*\n");
+			_glslCode = std::regex_replace(result, lineContinuation, PREPROC_LINE_CONTINUATION_DISABLER);
+		}
+		static void reenableDirectives(std::string& _glslCode)
+		{
+			std::regex lineContinuation(PREPROC_LINE_CONTINUATION_ENABLER);
+			auto result = std::regex_replace(_glslCode, lineContinuation, " \\");
+			std::regex glMacro(PREPROC_GL__ENABLER);
+			result = std::regex_replace(result, glMacro, " GL_");
```

Comment:
Anything glsl specific or specialized for a single language should not exist here in ISHadeCompiler